### DISCUSSION TITLE
Nullifier scheme

### DISCUSSION THREAD

**Discussion by bobbinth** [0xMiden/miden-vm#441] [Ideas]
In previous discussions I've mentioned that we use two databases to keep track of unconsumed notes: the Note DB and the Nullifier DB. However, I have not described previously how exactly these two work together to accomplish this.

At the high level, these databases do the following:
- Note DB keeps track of all notes ever created using an append-only accumulator (e.g., Merkle mountain range).
- Nullifier DB keeps track of consumed notes using a Sparse Merkle tree.

Leaves in the Note DB are not individual notes but rather Merkle trees such that each tree contains all notes created in a given block. Thus, a single leaf gets added to the Note DB with each new block. The note tree for each block itself contains trees of notes aggregated in a given transaction batch - but describing the exact structure of the Notes DB is not the main point of this article. The important part is that for every note ever created, we should be able to produce an inclusion path, and at the leaf of this path we will have a note hash.

A note hash is computed simply by hashing all components of a note. Specifically, components of a note are:

![image](https://user-images.githubusercontent.com/43513081/197080478-0ff445d4-ab8e-4ded-afe2-4200828c3441.png)

where:
- `vault_hash` is a commitment to all assets contained within a note.
- `script_root` is a root of a MAST for the note's spend script.
- `serial_num` is a value chosen randomly by the note's recipient.

Thus, a note hash can be computed as:
```
note_hash = hash(vault_hash, script_root, serial_num)
```

Where $hash$ is a native hash function used by the VM.

The above scheme is binding: we cannot change anything about the note without changing the hash. It is also hiding in that we need to know all 3 components of a note to be able to match a note to its hash. For example, knowing a note's `vault_hash` and `script_root` is not enough to figure if note hashes to a given value. We must also know the `serial_num`.

As mentioned above, note hashes are appended to the Note DB whenever new notes are created. However, when notes are consumed, we record note nullifiers in the Nullifier DB. We could have recorded note hashes in the Nullifier DB, but that would make notes linkable. Specifically, when a note is consumed (assuming it is consumed in a local transaction), we don't want to reveal which note was consumed exactly. Thus, we want the nullifier to have the following properties:

1. A nullifier should be derived from the note deterministically such that it is infeasible to derive two different nullifiers for the same note.
2. Given a nullifier and a note hash, it should not be possible to tell whether these were derived from the same note. Of course, if we know all components of a note, then we should be able to tell whether a nullifier for this note was computed correctly.

Originally, I was thinking that we could compute a nullifier simply as:
```
nullifier = hash(serial_num)
```
This would satisfy the two properties mentioned above (the nullifier is derived deterministically, and nullifier and note hash are unlinkable), and it might have worked in a fully private setting, where `serial_num` is known only to the note's recipient. But in a setting when we have private and public notes, this creates an attack vector.

For example, let's say we create a public note with some serial number $a$. This serial number can be observed by anyone on the network. A malicious user could create another note with the same serial number, and then consume it. Once this happens, an entry will be added to the Nullifier database indicating that a note with nullifier $hash(a)$ has been consumed. This would make it impossible for us to consume our original note.

To address this issue, we can compute the nullifier like so:
```
nullifier = hash(vault_hash, script_root, hash(serial_num))
```
This is similar to `note_hash` computation, but here we hash the `serial_num` one extra time. The properties we get from this are:

1. It is infeasible to derive two different nullifiers for the same note.
2. Given a nullifier and a note hash, there is no way to tell if they came from the same note.
3. If nullifiers for two notes are the same, then these two notes must be identical.

The last point means that it would make no sense for an attacker to create a note with the same serial number as to get the same nullifier they would need to create an identical note.

While I think the above scheme should work, I haven't spent much time analyzing it and would appreciate any feedback (especially if anyone sees holes in it).

**Comment by bobbinth** [comment]
Actually, the above description did not take into account some new elements described in #356. Specifically, new structure of a note would look like this:

![image](https://user-images.githubusercontent.com/43513081/197129997-a8856aca-aa6b-4f7e-8a1a-f8b3df5771c4.png)

where `block_ref` is the block reference number of a note (which must be smaller or equal to the current block height at the time of note creation).

The above changes our definition of note hash and nullifier computations as follows:
```
note_hash = hash(vault_hash, script_root, block_ref, serial_num)
nullifier = hash(vault_hash, script_root, block_ref, hash(serial_num))
```

**Reply by bobbinth** [reply]
I've realized that given the structure of Note DB, specifically that each leaf contains notes for a given block (as shown below), we actually don't need to put block reference number into the note itself. This number can be unambiguously determined any time we have an inclusion proof for a note.

![image](https://user-images.githubusercontent.com/43513081/197350677-4ed6c2cf-79da-47ea-ab48-dacacae1c599.png)

Thus, we can go back to the note structure described in the original post. Specifically:

![image](https://user-images.githubusercontent.com/43513081/197350835-73732014-23b7-4371-8bde-2bd9895707ee.png)

And then we can define note hash and nullifier computations as follows:
```
note_hash = hash(vault_hash, script_root, hash(serial_num))
nullifier = hash(vault_hash, script_root, serial_num)
```

Here, `serial_num` is hashed separately before being included into `note_hash` to improve privacy as discussed in https://github.com/maticnetwork/miden/discussions/441#discussioncomment-3936914. Specifically, this ensures that to compute a note hash, one does not need to know the serial number (but does need to know hash of the serial number).

**Reply by bobbinth** [reply]
It also may make sense to define `note_hash` as follows:
```
note_hash = hash(vault_hash, hash(script_root, hash(serial_num)))
```
It does require one extra hash, but it may simplify creating notes within a transaction kernel. Specifically, we could define a `recipient` value as follows:
```
recipient = hash(script_root, hash(serial_num))
``` 
And then we could have procedures like `create_note(asset, recipient)`.

**Comment by vlopes11** [comment]
I have a question about who sets the serial number. Given the following scenario

1.1) Let `C` be an account that is deployed and provide non-fungible assets functionalities
1.2) Let `A` be the owner of a given asset `x` in `C`
1.3) Let `B` be the owner of other assets of no relevance for this example, in `C`

Provided `A` wants to send assets to `B` by spending the asset `x` and creating another asset `y` with serial number `z`

2.1) `A` will compute locally a proof of correctness `p` by executing the `send` script of the account `C`
2.2) The proof `p` will bind the asset `x` to `y` via serial number `z`. Therefore, the serial number `z` must be known by `A` so he can craft the proof `p`.

Wouldn't `2.2` allow `A` to spend the asset `y` (because he knows the serial number `z`), when only `B` should be allowed to do so?

It would be safe, however, if we have a form of asymmetric cryptography as serial number, allowing `A` to generate `z` given a public identify of `B`, but not being able to spend it if the `send` expects a secret knowledge of the public identity of `B` in order to be successful. Is that the case?

**Reply by grjte** [reply]
I may be misunderstanding, but let me try to describe how I see this working in the context of our system.

`C` is an account in the global state, which means that transactions that update its state must be executed by the network. Therefore, `A` can locally prove a transaction that sends a note for `C` to consume, but `A` cannot locally consume that note against the account `C`. The transaction that consumes the note must be run by the network.

So for this example, what happens is something like the following:

- `A` proves a local transaction which outputs a note with serial number `z`. The script in that note calls `send` on the interface of `C` telling it to spend `x` and assign `y` to `B`.
- A network transaction consumes the note with serial number `z`. It runs `send` in `C`'s interface, which will check the predicates that `C` has implemented in its code (e.g. `x` can only be spent if this note came from `A`).
- `C`'s state is now updated; `x` is no longer assigned to `A` and `y` is assigned to `B`

> Wouldn't 2.2 allow A to spend the asset y (because he knows the serial number z), when only B should be allowed to do so?

So to address your question - `A` does not know a serial number for asset `y`, since `y` is contained in `C` which has implemented an interface with predicates. `A` only knows the serial number of the note it sends which gets consumed if the note script execution against `C` succeeds.

**Reply by bobbinth** [reply]
> Provided `A` wants to send assets to `B` by spending the asset `x` and creating another asset `y` with serial number `z`

If `A` wants to send asset $x$ to `B` (assuming `A` already holds $x$ in its vault), they don't need to spend $x$ and create $y$. The way this could work is that `A` executes a transaction against their account which creates note `N`, and then `B` executes another transaction which consumes note `N`.

The interesting part is how we build `N`. From the original post, a note consists of vault hash, script root, block reference number, and a serial number. In our case:

- The vault of the note would contain only a single asset $x$.
- The script root would be for a program which calls `receive()` method on `B`'s account. Let's denote this root as $p$.
- Block reference number could be chosen by either by `A` or `B` (more on this below). Let's denote it as $r$.
- Serial number could also be chosen by `A` or `B` as well. We'll denote as $z$ as before.

So, the hash of the note would be computed as follows:
```
note_hash = hash(x, p, r, z)
```
Thus, to create this note, `A` needs to know all 4 components. Depending on the interaction model, `A` could learn values for these in different ways.

First, let's consider a model where `B` sends a request for $x$ to `A`. This request could be sent out of band. The request would include all 4 components because:
- `A` needs to know $x$ and $p$ to know what to send and to whom.
- If $r$ and $z$ also come from `B` than `B` can compute note hash without any further interaction with `A` and, thus, can monitor the network (or delegate this monitoring to someone) for the note with expected hash to be created.

Second, let's consider the model where `A` sends $x$ to `B` without a request (e.g., by scanning a QR code). In this case:
- $x$, $r$, and $z$ would need to be set by `A` as they cannot be encoded in a QR code.
- $p$ could come from the QR code.

An open question is how would `B` know that `N` was sent to them? In case components of the note are public (i.e., stored on chain), `B` could scan for $r$. But in case of a private note (when only note hash is record on chain), $B$ would have no way of learning that `N` is addressed to them and actually won't be able to consume the notes as they won't know values for $x$, $r$, and $z$ - so, we'll need to come up with a mechanism to handle that.

Another consideration is that `A` may want to make $p$ conditional. For example, they may want to have a script root such that either $p$ gets executed, or after some time (e.g., 1 week) some other predicate $q$ can be executed to claim the note. This way, if `B` fails to consume the note within a week, `A` could claim it back.

The above makes me think that we may want to have some kind of "note templates" with well-defined format. We could start out with a handful of such templates and expand them over time.

#### Privacy improvement
In the first interaction model described above (`B` requests $x$ from `A`) privacy of `B` against `A` is somewhat limited. This is because if `A` knows $z$ they would be able to tell when note `N` gets consumed. It is not a big deal in this case as `A` knows who the note is being sent to in the first place, but there could be use cases where it is important to hide this information. We can improve on the original design by changing note hash computation as follows (but we keep nullifier computation similar to before):
```
note_hash = hash(vault_hash, script_root, hash(block_ref, serial_num))
nullifier = hash(vault_hash, script_root, block_ref, hash(serial_num))
```
This way, `B` needs to send the following pieces of info to `A`:
- $x$ - so that `A` could construct vault root.
- $p$ - so that `A` could verify that the script has the right format.
- $r$ - so that `A` could include it as a public input into the proof.
- $hash(r, z)$ - so that `A` could construct note hash without learning the value of $z$.

**Reply by bobbinth** [reply]
Overall, I feel like we should flash out potential interaction models a bit more. At the high level, it seems like we could have the following dimensions:

- in-band vs. out-of-band communication
- public vs. private notes

Approaches to how these could be handled could be very different - but I also wonder if we could find a unifying theme.

In all cases we want the note to be easily discoverable by the recipient, but we also don't want to reveal more information about note contents than necessary.

Let's take an example of out-of-band private notes (which I think is probably the simplest case to handle). If all communication happens out of band and `A` and `B` agree on the exact structure of the note (i.e., on note hash) before the note gets created, `B` knows exactly what to look for. There could be multiple ways `B` could "listen" for the note:

1. They could scan through all newly created notes themselves. This is probably going to be impractical for most users.
2. They could delegate scanning to someone else and there are multiple options here too:
   a. They could give the exact hash of the note they are expecting to some 3rd party.
   b. They could provide some info about the hash which would enable some kind of "fuzzy matching" via Bloom filters or something else.
  c. They could also manipulate hash of the note itself by grinding on the note's serial number. For example, they could try many different serial numbers until hash of the note has a specific 20-bit prefix. Then, when delegating scanning, they could ask for all notes with this 20-bit prefix.

The above is just brainstorming, and there could be better ways of achieving our goals.

**Reply by jdkanani** [reply]
*Possible that this comment might not be relevant here. Also, I am still trying to understand the Miden system. This mechanism may not work at all. Just consider it a comment from a noob.*

I might be missing many discussions around this, but I was still thinking about how a simple transfer or swap would look.  Here are a few thoughts -

- Recording a note on the chain and consuming it in the second tx would be the easiest way to implement the flow, but each transfer note has to include the expiry date in case the receiver doesn't claim it. Notifying the receiver and making them sign the tx in a limited period is not that intuitive. If the note transfer expires, the sender has to claim it back. Until that, the amount gets locked in between. The sender might have to re-call and send a new note again in the majority of the scenario. Not so good UX for transfer/swap.
- Sender `A` can create a note off-chain using the serial number provided by receiver `B`, sign the transaction, create the proof, and give it to `B` to consume it on-chain. Once `B` receives it, it can verify and submits the consumption tx with the note creation tx.
- Node script should include short expiry and recall if `B` tries to delay the transfer. But here, `A` shouldn't have to wait longer and doesn't have to recall if `B` is honest.
- Transfer/Swap must only be considered final once it is recorded on the chain.

Possible that this might break the off-chain proof creation system the way I am not able to think yet.

**Reply by bobbinth** [reply]
I'll write up a more detailed description of how I think this will work (including some thoughts I described in https://github.com/0xPolygonMiden/miden-base/discussions/7#discussioncomment-4489919), but here are a couple of quick comments.

> Sender `A` can create a note off-chain using the serial number provided by receiver `B`, sign the transaction, create the proof, and give it to `B` to consume it on-chain. Once `B` receives it, it can verify and submits the consumption tx with the note creation tx.

Assuming I understood this correctly, I am not sure this is preferable to the first option. If `A` gives gives a transaction to `B` without putting it on chain, then in the period of time between `A` giving this transaction to `B` and `B` executing it, `A` could execute a different transaction agains his account, and make the transaction he gave to `B` un-executable.

> Recording a note on the chain and consuming it in the second tx would be the easiest way to implement the flow, but each transfer note has to include the expiry date in case the receiver doesn't claim it. Notifying the receiver and making them sign the tx in a limited period is not that intuitive.

I actually think this may not be so bad. First, in my mind, the "recall" option is there only to prevent "catastrophic" situations (e.g., sending assets to non-existent address). So, the non-recallable period could be quite long. I've always thought about it as being on the order of several months. Basically, if I screwed up somehow and the intended recipient can't receive the assets for some reason, I know I can get the assets back within a few months.

My other assumption is that vast majority of notes will be consumed in a span of a month or so. Ideally, note consumption will be either semi or fully automated at the wallet level. For example, once the wallet software detects an incoming note, it would create a transaction to consume this note automatically (or maybe present a confirmation option to the user).

Finally, we can always default to non-recallable notes. These notes don't have the same protection against screw-ups, but they give the same "finality" as regular crypto transactions (i.e., assuming `A` didn't screw up, `B` can consume such a note at any time - even a year later).

**Reply by bobbinth** [reply]
Added a detailed description of how user-to-user transfers could work [here](https://github.com/0xPolygonMiden/miden-base/discussions/11).

**Reply by jdkanani** [reply]
My comment was more around swap where I want to execute the swap and makes it invalid after sometime if it is not executed. Let me think more and get back.

**Reply by jdkanani** [reply]
One comment — 

> Assuming I understood this correctly, I am not sure this is preferable to the first option. If A gives gives a transaction to B without putting it on chain, then in the period of time between A giving this transaction to B and B executing it, A could execute a different transaction agains his account, and make the transaction he gave to B un-executable.

Yes. That’s true. It means that B considers this as an invalid transfer. B should only assume transfer is finalized once it is recorded on the chain. Goal is not to make transfer/swap without executing on-chain but make full-transfer/swap atomic (or rather in one tx) instead of locking amount on-chain.
